src/common/CommandTable.h

```diff
@@ -0,0 +1,91 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software 
+ * Foundation.  See file COPYING.
+ * 
+ */
+
+#ifndef COMMAND_TABLE_H_
+#define COMMAND_TABLE_H_
+
+#include "messages/MCommand.h"
+
+class CommandOp
+{
+  public:
+  ConnectionRef con;
+  ceph_tid_t tid;
+
+  std::vector<std::string> cmd;
+  bufferlist    inbl;
+  Context      *on_finish;
+  bufferlist   *outbl;
+  std::string  *outs;
+
+  MCommand *get_message(const uuid_d &fsid) const
+  {
+    MCommand *m = new MCommand(fsid);
+    m->cmd = cmd;
+    m->set_data(inbl);
+    m->set_tid(tid);
+
+    return m;
+  }
+
+  CommandOp(const ceph_tid_t t) : tid(t) {}
+};
+
+template<typename T>
+class CommandTable
+{
+protected:
+  ceph_tid_t last_tid;
+  std::map<ceph_tid_t, T*> commands;
```

Comment:
If someone misuses this interface, it'll just leak the pointers.  That's not great.  I'd like get_commands() and get_command() (since they give up ownership of the pointer) to remove the returned pointers from the map.  I'd further like the destructor for CommandTable to assert that the map is empty.